File: informes/Porcentaje_llenado_peorescasos/generar_pdf_contenedores.R
# Genera un PDF con el detalle por gid del resultado de analizar_peores_contenedores()
# (ver PorcentajeLlenado.R): identificación, historial crudo filtrado a ese gid,
# y sus métricas agregadas. Un PDF único con una sección por gid.
# Sigue el mismo patrón de entorno virtual de Python que funciones_utiles.R.

# ── PDF por contenedor ─────────────────────────────────────────────────────────
#
# Parámetros:
#   resultado_analisis - lista devuelta por analizar_peores_contenedores()
#                         (debe traer $resultado y $datos_filtrados).
#   nombre_salida       - nombre del archivo .pdf de salida (se guarda en
#                          informes/salidas/). Si es NULL, se arma a partir de
#                          $descripcion_filtros del resultado_analisis.
#   instalar_librerias  - si TRUE, instala/verifica pandas, pyreadr y reportlab
#                          en el entorno virtual antes de correr (solo hace
#                          falta la primera vez o al actualizar dependencias).
#
# Uso:
#   ver <- analizar_peores_contenedores(gids = c("184421", "184471", "107365"))
#   generar_pdf_contenedores(ver)
generar_pdf_contenedores <- function(resultado_analisis, nombre_salida = NULL, instalar_librerias = FALSE) {
  library(reticulate)

  if (is.null(resultado_analisis$resultado) || is.null(resultado_analisis$datos_filtrados)) {
    stop("resultado_analisis debe ser la lista devuelta por analizar_peores_contenedores() (con $resultado y $datos_filtrados).")
  }

  # 1. Verificar que el entorno virtual de Python existe; crearlo si no (mismo patrón que funciones_utiles.R)
  venv_path <- file.path(virtualenv_root(), "r-reticulate")
  if (!virtualenv_exists(venv_path)) {
    message("El entorno virtual no existe en: ", venv_path, ". Intentando crearlo...")
    tryCatch(
      {
        virtualenv_create("r-reticulate")
      },
      error = function(e) {
        message("No se encontró Python instalarlo automáticamente...")
        install_python()
        virtualenv_create("r-reticulate")
      }
    )
  }
  use_virtualenv("r-reticulate", required = TRUE)

  # 2. Instalar librerías Python si se solicitó (solo instala las que falten)
  if (instalar_librerias) {
    message("Verificando librerías de Python instaladas...")
    librerias <- c("pandas", "pyreadr", "reportlab")
    faltantes <- c()
    for (lib in librerias) {
      if (!py_module_available(lib)) {
        faltantes <- c(faltantes, lib)
      }
    }
    if (length(faltantes) > 0) {
      message("Instalando librerías de Python faltantes: ", paste(faltantes, collapse = ", "))
      py_install(faltantes)
    } else {
      message("Todas las librerías de Python ya están instaladas.")
    }
  }

  # 3. Guardar los datos para que Python los lea con pyreadr (un .rds por data.frame,
  #    igual que el resto de los generadores de PDF del proyecto)
  dir_datos <- "informes/Porcentaje_llenado_peorescasos/data"
  dir.create(dir_datos, showWarnings = FALSE, recursive = TRUE)
  saveRDS(resultado_analisis$resultado, file.path(dir_datos, "pdf_resultado.rds"))
  saveRDS(resultado_analisis$datos_filtrados, file.path(dir_datos, "pdf_crudo.rds"))

  # 4. Nombre de salida: el que se pase, o uno armado a partir de los filtros del análisis
  if (is.null(nombre_salida)) {
    descripcion <- if (!is.null(resultado_analisis$descripcion_filtros)) resultado_analisis$descripcion_filtros else "contenedores"
    nombre_salida <- paste0("informe_contenedores_", descripcion, ".pdf")
  }

  # 5. Pasar el nombre de salida a Python como variable de entorno y correr el script
  Sys.setenv(PDF_CONTENEDORES_SALIDA = nombre_salida)
  message("Generando PDF...")
  py_run_file("informes/Porcentaje_llenado_peorescasos/generar_pdf_contenedores.py")
  Sys.unsetenv("PDF_CONTENEDORES_SALIDA")
  message("Generación de PDF completada.")
}

# Uso de referencia:
# ver <- analizar_peores_contenedores(gids = c("184421", "184471", "107365"))
# generar_pdf_contenedores(ver)

File: informes/Porcentaje_llenado_peorescasos/generar_pdf_contenedores.py
import os
import pyreadr
import pandas as pd
from reportlab.lib import colors
from reportlab.lib.pagesizes import A4
from reportlab.lib.units import cm
from reportlab.platypus import SimpleDocTemplate, Table, TableStyle, Paragraph, Spacer, PageBreak
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle

# Genera un PDF con una sección por cada gid presente en $resultado (salida de
# analizar_peores_contenedores() en PorcentajeLlenado.R): identificación,
# historial crudo de visitas filtrado a ese gid, y sus métricas agregadas.
# Se invoca vía generar_pdf_contenedores.R (reticulate), que deja los datos
# en data/pdf_resultado.rds y data/pdf_crudo.rds antes de correr este script.

BASE_DIR = os.path.dirname(os.path.abspath(__file__))
DATA_DIR = os.path.join(BASE_DIR, "data")
SALIDAS_DIR = os.path.normpath(os.path.join(BASE_DIR, "..", "salidas"))

COLUMNAS_IDENTIFICACION = ["Circuito", "Posicion", "Direccion", "gid", "Circuito_corto", "Oficina"]
COLUMNAS_CRUDO = ["Fecha", "Levantado", "Turno_levantado", "Porcentaje_llenado", "Incidencia", "Condicion"]
COLUMNAS_RESULTADO = ["n_recolecciones", "n_saturaciones", "tasa_saturacion", "llenado_promedio", "frecuencia_dias", "cuadrante"]

ESTILO_TABLA = TableStyle([
    ("BACKGROUND", (0, 0), (-1, 0), colors.whitesmoke),
    ("LINEBELOW", (0, 0), (-1, 0), 1, colors.grey),
    ("GRID", (0, 0), (-1, -1), 0.5, colors.lightgrey),
    ("VALIGN", (0, 0), (-1, -1), "TOP"),
    ("TOPPADDING", (0, 0), (-1, -1), 3),
    ("BOTTOMPADDING", (0, 0), (-1, -1), 3),
    ("ROWBACKGROUNDS", (0, 1), (-1, -1), [colors.white, colors.HexColor("#f7f7f5")]),
])

ESTILO_HEADER = ParagraphStyle("Header", fontName="Helvetica-Bold", fontSize=7.5, leading=9)
ESTILO_CELDA = ParagraphStyle("Celda", fontName="Helvetica", fontSize=7.5, leading=9)

ANCHO_TABLA = 18 * cm
# Peso relativo de cada columna conocida (columnas no listadas usan 1.5 por defecto);
# cuadrante es texto largo y necesita mucho más espacio que el resto para no cortarse.
PESO_COLUMNA = {
    "Direccion": 2.6, "Incidencia": 2, "Condicion": 2, "cuadrante": 4,
    "Circuito": 2.2, "Circuito_corto": 1.3, "Turno_levantado": 1.4,
}


def df_a_tabla(df, columnas, ancho_total=ANCHO_TABLA):
    columnas_presentes = [c for c in columnas if c in df.columns]
    # astype(object) primero: pyreadr trae columnas de factor de R como Categorical,
    # que no admite fillna("") si "" no es una categoría existente. where() en vez de
    # fillna() evita el warning de downcasting de pandas al mezclar tipos.
    df_fmt = df[columnas_presentes].astype(object)
    df_fmt = df_fmt.where(df_fmt.notna(), "").astype(str)

    encabezado = [Paragraph(c, ESTILO_HEADER) for c in columnas_presentes]
    filas = [[Paragraph(valor, ESTILO_CELDA) for valor in fila] for fila in df_fmt.values.tolist()]
    datos = [encabezado] + filas

    pesos = [PESO_COLUMNA.get(c, 1.5) for c in columnas_presentes]
    total_peso = sum(pesos)
    anchos = [ancho_total * p / total_peso for p in pesos]

    t = Table(datos, colWidths=anchos)
    t.setStyle(ESTILO_TABLA)
    return t


def formatear_resultado(fila):
    fila = fila.copy()
    if "tasa_saturacion" in fila:
        fila["tasa_saturacion"] = f"{float(fila['tasa_saturacion']) * 100:.1f}%"
    if "llenado_promedio" in fila:
        fila["llenado_promedio"] = f"{float(fila['llenado_promedio']):.1f}"
    if "frecuencia_dias" in fila:
        fila["frecuencia_dias"] = f"{float(fila['frecuencia_dias']):.2f}"
    return fila


def generar_pdf(resultado, crudo, ruta_salida):
    doc = SimpleDocTemplate(
        ruta_salida, pagesize=A4,
        rightMargin=1.5 * cm, leftMargin=1.5 * cm,
        topMargin=1.5 * cm, bottomMargin=1.5 * cm
    )
    estilos = getSampleStyleSheet()
    estilo_titulo = ParagraphStyle("Titulo", parent=estilos["Heading1"], fontSize=13, spaceAfter=10)
    estilo_gid = ParagraphStyle("Gid", parent=estilos["Heading2"], fontSize=12, spaceBefore=4, spaceAfter=6, textColor=colors.HexColor("#a50f15"))
    estilo_seccion = ParagraphStyle("Seccion", parent=estilos["Heading3"], fontSize=10, spaceBefore=8, spaceAfter=4)

    elementos = [Paragraph("Informe por contenedor — peores casos de saturación", estilo_titulo)]

    gids = resultado["gid"].astype(str).tolist()
    for i, gid in enumerate(gids):
        fila_resultado = resultado[resultado["gid"].astype(str) == gid].iloc[0]
        crudo_gid = crudo[crudo["gid"].astype(str) == gid].copy()
        crudo_gid["Fecha"] = crudo_gid["Fecha"].astype(str)
        crudo_gid = crudo_gid.sort_values(by="Fecha", ascending=False)

        elementos.append(Paragraph(f"Contenedor {gid}", estilo_gid))

        # --- Identificación: fila más reciente del crudo para este gid ---
        if not crudo_gid.empty:
            identificacion = crudo_gid.iloc[[0]]
            elementos.append(Paragraph("Identificación", estilo_seccion))
            elementos.append(df_a_tabla(identificacion, COLUMNAS_IDENTIFICACION))
            elementos.append(Spacer(1, 8))

        # --- Historial crudo filtrado a este gid ---
        elementos.append(Paragraph(f"Historial de visitas ({len(crudo_gid)} filas)", estilo_seccion))
        if crudo_gid.empty:
            elementos.append(Paragraph("Sin filas crudas para este gid en la ventana analizada.", estilos["Normal"]))
        else:
            elementos.append(df_a_tabla(crudo_gid, COLUMNAS_CRUDO))
        elementos.append(Spacer(1, 8))

        # --- Resultado agregado ---
        elementos.append(Paragraph("Resultado (métricas agregadas)", estilo_seccion))
        fila_fmt = formatear_resultado(fila_resultado)
        tabla_resultado = pd.DataFrame([fila_fmt])
        elementos.append(df_a_tabla(tabla_resultado, COLUMNAS_RESULTADO))

        if i < len(gids) - 1:
            elementos.append(PageBreak())

    doc.build(elementos)


if __name__ == "__main__":
    resultado = pyreadr.read_r(os.path.join(DATA_DIR, "pdf_resultado.rds"))[None]
    crudo = pyreadr.read_r(os.path.join(DATA_DIR, "pdf_crudo.rds"))[None]

    if resultado.empty:
        print("No hay contenedores en el resultado; no se genera PDF.")
    else:
        os.makedirs(SALIDAS_DIR, exist_ok=True)
        nombre_salida = os.environ.get("PDF_CONTENEDORES_SALIDA", "informe_contenedores.pdf")
        ruta_salida = os.path.join(SALIDAS_DIR, nombre_salida)
        generar_pdf(resultado, crudo, ruta_salida)
        print(f"PDF generado: {ruta_salida}")
